### ISSUE TITLE
Allow selecting source of session data for reports (DiscoveryService vs FederatedLoginEvent)

### ISSUE THREAD

**Issue by vladimir-mencl-eresearch** [ausaccessfed/reporting-service#176]
Hi @smangelsdorf 

As discussed - I've started implementing support for selecting the session data source for reports that use login session - with the source being either DiscoveryService or FederatedLoginEvent.

I'm creating this issue as a discussion space before I create a PR to give you a chance to shape where I'm heading earlier on.

Cheers,
Vlad

**Comment by vladimir-mencl-eresearch** [comment]
Hi @smangelsdorf 

I have the first cut.

It does what I expect it to do.

I've added the source selection as an extra field in `app/views/administrator_reports/_form.html.slim` and `app/views/shared/_subscriber_report_form.html.slim`.

This hooks almost exactly into the reports that need this field - I only need to hide this from the FederationGrowthReport.  (Either I find a way to make the field conditional, or I'll have to clone the template into two versions).

Or, I may instead create a template with just the select field and include that into the report templates that need this.

And I may also extract the list of choices from the template into a constant defined in a Ruby file - your preference on that?

I'll paste my current full git diff into the next message - please let me know how close I am to your target style - and where you'd like me to make adjustments.

I'd add tests once I know I'm on the right track.

Cheers,
Vlad

**Comment by vladimir-mencl-eresearch** [comment]
My current `git diff`

````
diff --git a/app/controllers/administrator_reports_controller.rb b/app/controllers/administrator_reports_controller.rb
index 261be3c..69e6de9 100644
--- a/app/controllers/administrator_reports_controller.rb
+++ b/app/controllers/administrator_reports_controller.rb
@@ -4,6 +4,9 @@ class AdministratorReportsController < ApplicationController
   before_action { check_access! 'admin:report' }
   before_action :set_range_params,
                 except: %i[subscriber_registrations_report index]
+  before_action :set_source,
+                except: %i[subscriber_registrations_report
+                           federation_growth_report index]

   def index; end

@@ -25,28 +28,28 @@ class AdministratorReportsController < ApplicationController
   def daily_demand_report
     return if params[:start].blank? || params[:end].blank?

-    report = DailyDemandReport.new(start, finish)
+    report = DailyDemandReport.new(start, finish, @source)
     @data = JSON.generate(report.generate)
   end

   def federated_sessions_report
     return if params[:start].blank? || params[:end].blank?

-    report = FederatedSessionsReport.new(start, finish, scaled_steps)
+    report = FederatedSessionsReport.new(start, finish, scaled_steps, @source)
     @data = JSON.generate(report.generate)
   end

   def identity_provider_utilization_report
     return if params[:start].blank? || params[:end].blank?

-    report = IdentityProviderUtilizationReport.new(start, finish)
+    report = IdentityProviderUtilizationReport.new(start, finish, @source)
     @data = JSON.generate(report.generate)
   end

   def service_provider_utilization_report
     return if params[:start].blank? || params[:end].blank?

-    report = ServiceProviderUtilizationReport.new(start, finish)
+    report = ServiceProviderUtilizationReport.new(start, finish, @source)
     @data = JSON.generate(report.generate)
   end

@@ -57,6 +60,10 @@ class AdministratorReportsController < ApplicationController
     @end = params[:end]
   end

+  def set_source
+    @source = params[:source]
+  end
+
   def start
     return nil if params[:start].blank?
     Time.zone.parse(params[:start]).beginning_of_day
diff --git a/app/controllers/federation_reports_controller.rb b/app/controllers/federation_reports_controller.rb
index 217625f..f315e0e 100644
--- a/app/controllers/federation_reports_controller.rb
+++ b/app/controllers/federation_reports_controller.rb
@@ -2,6 +2,7 @@

 class FederationReportsController < ApplicationController
   before_action :set_range
+  before_action :set_source

   def federation_growth_report
     public_action
@@ -16,7 +17,7 @@ class FederationReportsController < ApplicationController
     public_action

     @data = Rails.cache.fetch('public/federated-sessions') do
-      report = FederatedSessionsReport.new(@start, @end, 10)
+      report = FederatedSessionsReport.new(@start, @end, 10, @source)
       JSON.generate(report.generate)
     end
   end
@@ -25,7 +26,7 @@ class FederationReportsController < ApplicationController
     public_action

     @data = Rails.cache.fetch('public/daily-demand') do
-      report = DailyDemandReport.new(@start, @end)
+      report = DailyDemandReport.new(@start, @end, @source)
       JSON.generate(report.generate)
     end
   end
@@ -36,4 +37,8 @@ class FederationReportsController < ApplicationController
     @start = 1.year.ago.beginning_of_day
     @end = Time.zone.tomorrow.beginning_of_day
   end
+
+  def set_source
+    @source = params[:source]
+  end
 end
diff --git a/app/controllers/subscriber_reports.rb b/app/controllers/subscriber_reports.rb
index 3e971d7..a505c9b 100644
--- a/app/controllers/subscriber_reports.rb
+++ b/app/controllers/subscriber_reports.rb
@@ -4,6 +4,7 @@ class SubscriberReports < ApplicationController
   before_action { permitted_objects(model_object) }
   before_action :requested_entity
   before_action :set_range_params
+  before_action :set_source
   before_action :access_method

   private
@@ -24,8 +25,8 @@ class SubscriberReports < ApplicationController
   def generate_report(report_type, steps = nil)
     @entity_id = params[:entity_id]

-    return report_type.new(@entity_id, start, finish, steps) if steps
-    report_type.new(@entity_id, start, finish)
+    return report_type.new(@entity_id, start, finish, steps, @source) if steps
+    report_type.new(@entity_id, start, finish, @source)
   end

   def access_method
@@ -50,6 +51,10 @@ class SubscriberReports < ApplicationController
     @end = params[:end]
   end

+  def set_source
+    @source = params[:source]
+  end
+
   def start
     return nil if params[:start].blank?
     Time.zone.parse(params[:start]).beginning_of_day
diff --git a/app/jobs/process_incoming_f_ticks_events.rb b/app/jobs/process_incoming_f_ticks_events.rb
index df4f902..0c350e9 100644
--- a/app/jobs/process_incoming_f_ticks_events.rb
+++ b/app/jobs/process_incoming_f_ticks_events.rb
@@ -18,6 +18,7 @@ class ProcessIncomingFTicksEvents

   def incoming_events
     IncomingFTicksEvent
-      .where('discarded != ? AND created_at <= ?', true, Time.zone.now)
+      .where('discarded != ? AND created_at <= ?', true, Time.zone.now -
+        86_400 * ENV['SKIP_RECENT_DAYS'].to_i)
   end
 end
diff --git a/app/models/federated_login_event.rb b/app/models/federated_login_event.rb
index 00309d6..15945e5 100644
--- a/app/models/federated_login_event.rb
+++ b/app/models/federated_login_event.rb
@@ -3,6 +3,24 @@
 class FederatedLoginEvent < ActiveRecord::Base
   valhammer

+  belongs_to :identity_provider,
+             foreign_key: :asserting_party,
+             primary_key: :entity_id
+
+  belongs_to :service_provider,
+             foreign_key: :relying_party,
+             primary_key: :entity_id
+
+  alias_attribute :selected_idp, :asserting_party
+  alias_attribute :initiating_sp, :relying_party
+
+  scope(:within_range, lambda { |start, finish|
+    where(arel_table[:timestamp].gteq(start)
+      .and(arel_table[:timestamp].lteq(finish)))
+  })
+
+  scope(:sessions, -> { where(result: 'OK') })
+
   def create_instance(event)
     data = fields(event.data)
     update login_event_hash(data)
diff --git a/app/reports/daily_demand_report.rb b/app/reports/daily_demand_report.rb
index 247aac8..0dc9193 100644
--- a/app/reports/daily_demand_report.rb
+++ b/app/reports/daily_demand_report.rb
@@ -8,10 +8,11 @@ class DailyDemandReport < TimeSeriesReport
   units ''
   series sessions: 'Sessions'

-  def initialize(start, finish)
+  def initialize(start, finish, source)
     title = 'Daily Demand'
     @start = start
     @finish = finish
+    @source = source

     super(title, start: @start, end: @finish)
   end
diff --git a/app/reports/federated_sessions_report.rb b/app/reports/federated_sessions_report.rb
index 1e3ed41..c4aa66a 100644
--- a/app/reports/federated_sessions_report.rb
+++ b/app/reports/federated_sessions_report.rb
@@ -8,11 +8,12 @@ class FederatedSessionsReport < TimeSeriesReport
   units ''
   series sessions: 'Sessions'

-  def initialize(start, finish, steps)
+  def initialize(start, finish, steps, source)
     title = 'Federated Sessions'
     @start = start
     @finish = finish
     @steps = steps
+    @source = source

     super(title, start: @start, end: @finish)
   end
diff --git a/app/reports/identity_provider_daily_demand_report.rb b/app/reports/identity_provider_daily_demand_report.rb
index 71cfdcc..568c732 100644
--- a/app/reports/identity_provider_daily_demand_report.rb
+++ b/app/reports/identity_provider_daily_demand_report.rb
@@ -8,11 +8,12 @@ class IdentityProviderDailyDemandReport < TimeSeriesReport
   units ''
   series sessions: 'Sessions'

-  def initialize(entity_id, start, finish)
+  def initialize(entity_id, start, finish, source)
     @identity_provider = IdentityProvider.find_by(entity_id: entity_id)
     title = "IdP Daily Demand Report for #{@identity_provider.name}"
     @start = start
     @finish = finish
+    @source = source

     super(title, start: @start, end: @finish)
   end
diff --git a/app/reports/identity_provider_destination_services_report.rb b/app/reports/identity_provider_destination_services_report.rb
index 6864490..3cc85d1 100644
--- a/app/reports/identity_provider_destination_services_report.rb
+++ b/app/reports/identity_provider_destination_services_report.rb
@@ -7,11 +7,12 @@ class IdentityProviderDestinationServicesReport < TabularReport
   header ['SP Name', 'Total']
   footer

-  def initialize(entity_id, start, finish)
+  def initialize(entity_id, start, finish, source)
     @identity_provider = IdentityProvider.find_by(entity_id: entity_id)
     title = "IdP Destination Report for #{@identity_provider.name}"
     @start = start
     @finish = finish
+    @source = source

     super(title)
   end
diff --git a/app/reports/identity_provider_sessions_report.rb b/app/reports/identity_provider_sessions_report.rb
index cdfa6de..eae9418 100644
--- a/app/reports/identity_provider_sessions_report.rb
+++ b/app/reports/identity_provider_sessions_report.rb
@@ -8,12 +8,13 @@ class IdentityProviderSessionsReport < TimeSeriesReport
   units ''
   series sessions: 'Sessions'

-  def initialize(entity_id, start, finish, steps)
+  def initialize(entity_id, start, finish, steps, source)
     @identity_provider = IdentityProvider.find_by(entity_id: entity_id)
     title = "Identity Provider Sessions for #{@identity_provider.name}"
     @start = start
     @finish = finish
     @steps = steps
+    @source = source

     super(title, start: @start, end: @finish)
   end
diff --git a/app/reports/identity_provider_utilization_report.rb b/app/reports/identity_provider_utilization_report.rb
index 0e0c21c..c0cb0b1 100644
--- a/app/reports/identity_provider_utilization_report.rb
+++ b/app/reports/identity_provider_utilization_report.rb
@@ -8,10 +8,11 @@ class IdentityProviderUtilizationReport < TabularReport
   header %w[Name Sessions]
   footer

-  def initialize(start, finish)
+  def initialize(start, finish, source)
     title = 'Identity Provider Utilization Report'
     @start = start
     @finish = finish
+    @source = source

     super(title)
   end
diff --git a/app/reports/reports_shared_methods.rb b/app/reports/reports_shared_methods.rb
index fabaff0..4ccadd3 100644
--- a/app/reports/reports_shared_methods.rb
+++ b/app/reports/reports_shared_methods.rb
@@ -48,8 +48,15 @@ module ReportsSharedMethods
     report[time] ? (report[time].to_f / divider).round(decimal_places) : 0.0
   end

+  SESSION_SOURCES = {
+    'DS' => DiscoveryServiceEvent,
+    'IdP' => FederatedLoginEvent
+  }.freeze
+
   def sessions(where_args = {})
-    DiscoveryServiceEvent
+    # FederatedLoginEvent
+    # DiscoveryServiceEvent
+    SESSION_SOURCES[@source]
       .within_range(@start, @finish).where(where_args).sessions
   end

diff --git a/app/reports/service_provider_daily_demand_report.rb b/app/reports/service_provider_daily_demand_report.rb
index 9a136d8..fa87d28 100644
--- a/app/reports/service_provider_daily_demand_report.rb
+++ b/app/reports/service_provider_daily_demand_report.rb
@@ -8,11 +8,12 @@ class ServiceProviderDailyDemandReport < TimeSeriesReport
   units ''
   series sessions: 'Sessions'

-  def initialize(entity_id, start, finish)
+  def initialize(entity_id, start, finish, source)
     @service_provider = ServiceProvider.find_by(entity_id: entity_id)
     title = "SP Daily Demand Report for #{@service_provider.name}"
     @start = start
     @finish = finish
+    @source = source

     super(title, start: @start, end: @finish)
   end
diff --git a/app/reports/service_provider_sessions_report.rb b/app/reports/service_provider_sessions_report.rb
index 7168115..1213505 100644
--- a/app/reports/service_provider_sessions_report.rb
+++ b/app/reports/service_provider_sessions_report.rb
@@ -8,12 +8,13 @@ class ServiceProviderSessionsReport < TimeSeriesReport
   units ''
   series sessions: 'Sessions'

-  def initialize(entity_id, start, finish, steps)
+  def initialize(entity_id, start, finish, steps, source)
     @service_provider = ServiceProvider.find_by(entity_id: entity_id)
     title = "Service Provider Sessions for #{@service_provider.name}"
     @start = start
     @finish = finish
     @steps = steps
+    @source = source

     super(title, start: @start, end: @finish)
   end
diff --git a/app/reports/service_provider_source_identity_providers_report.rb b/app/reports/service_provider_source_identity_providers_report.rb
index c8e8471..abb560b 100644
--- a/app/reports/service_provider_source_identity_providers_report.rb
+++ b/app/reports/service_provider_source_identity_providers_report.rb
@@ -7,11 +7,12 @@ class ServiceProviderSourceIdentityProvidersReport < TabularReport
   header ['IdP Name', 'Total']
   footer

-  def initialize(entity_id, start, finish)
+  def initialize(entity_id, start, finish, source)
     @service_provider = ServiceProvider.find_by(entity_id: entity_id)
     title = "SP Source Identity Providers Report for #{@service_provider.name}"
     @start = start
     @finish = finish
+    @source = source

     super(title)
   end
diff --git a/app/reports/service_provider_utilization_report.rb b/app/reports/service_provider_utilization_report.rb
index b6251a1..2c57a5b 100644
--- a/app/reports/service_provider_utilization_report.rb
+++ b/app/reports/service_provider_utilization_report.rb
@@ -8,10 +8,11 @@ class ServiceProviderUtilizationReport < TabularReport
   header %w[Name Sessions]
   footer

-  def initialize(start, finish)
+  def initialize(start, finish, source)
     title = 'Service Provider Utilization Report'
     @start = start
     @finish = finish
+    @source = source

     super(title)
   end
diff --git a/app/views/administrator_reports/_form.html.slim b/app/views/administrator_reports/_form.html.slim
index 0fd44fb..d599960 100644
--- a/app/views/administrator_reports/_form.html.slim
+++ b/app/views/administrator_reports/_form.html.slim
@@ -14,6 +14,16 @@

     = date_field_tag :end, value = end_date, placeholder: 'click here to pick a date'

+  = field_block do
+    = label_tag(:source) do
+      | session data source
+
+
+
+
+
+    = select_tag :source, options_for_select([[ "Discovery Service", "DS"], [ "IdP event log", "IdP"] ], "DS"), placeholder: 'select the data source'
+
   = button_tag(type: 'submit', class: 'btn-lg btn-primary') do
     => icon_tag('ok')
     | Generate
diff --git a/app/views/shared/_subscriber_report_form.html.slim b/app/views/shared/_subscriber_report_form.html.slim
index 8852622..3a28fe9 100644
--- a/app/views/shared/_subscriber_report_form.html.slim
+++ b/app/views/shared/_subscriber_report_form.html.slim
@@ -24,6 +24,12 @@

       = date_field_tag :end, value = end_date, placeholder: 'click here to pick a date'

+    = field_block do
+      = label_tag(:source) do
+       | session data source
+
+      = select_tag :source, options_for_select([[ "Discovery Service", "DS"], [ "IdP event log", "IdP"] ], "DS"), placeholder: 'select the data source'
+
     = button_tag(type: 'submit', class: 'btn-lg btn-primary') do
       => icon_tag('ok')
       | Generate
diff --git a/config/dev_unicorn.rb b/config/dev_unicorn.rb
index 8e87974..8f028b6 100644
--- a/config/dev_unicorn.rb
+++ b/config/dev_unicorn.rb
@@ -1,8 +1,10 @@
 # frozen_string_literal: true

-listen 8080
+listen 8083
 preload_app false
 stdout_path '/dev/null'
+# stdout_path 'log/unicorn-stdout.log'
+stderr_path 'log/unicorn-stderr.log'

 root = File.expand_path(File.join(File.dirname(__FILE__), '..'))
 pid File.join(root, 'tmp', 'pids', 'unicorn.pid')
@@ -25,6 +27,11 @@ before_fork do |_, _|
   end
 end

+# set timeout to 10 minutes
+# timeout 600
+# set timeout to 8 hours
+timeout 28_800
+
 logger_obj = Logger.new($stderr)
 logger_obj.level = Logger::WARN
 logger logger_obj
````

**Comment by vladimir-mencl-eresearch** [comment]
I should add, part of what I'm doing is making a FederatedLoginEvent look like a DiscoveryServiceEvent - that's the aliases I'm adding in the model.

And of course, I will not be committing my changes to `config/dev_unicorn.rb` ... and I'll clean up whitespace in the templates.

How's this looking?

**Comment by smangelsdorf** [comment]
:+1:

Looks like it's off to a very sensible start. I'd prefer a version without the attribute aliases, if you can find another way to avoid the duplication. At a bit of a glance over the codebase, the attributes are only used in `ReportsSharedMethods`, so that might be the place to do it.

Nitpick: I'd like to see the form default to DS rather than forcing the user to select. I'm assuming that's how it works currently, because that's how the diff reads.

**Comment by vladimir-mencl-eresearch** [comment]
Thanks!

I'm now adjusting tests to make the code pass - and I'm finding little omissions in the main code while doing that.

Key questions I have as of now are:
* I'll have to deal with subscribed reports: I'll have to pass the selection from the report form into the subscription - so I'm expecting this will result into an extra field in the automated reports and so also a DB migration.
* In the tests, I have to make the tests pass a selection source: would you be fine with this being a hard-coded 'DS' ?  I can also add additional tests that would specifically test that test data from federated_login_events is passed through when source is IdP.

I'll have a look at what can be done to avoid the aliases...

Cheers,
Vlad

**Comment by smangelsdorf** [comment]
Makes sense, re: subscribed reports. I'll wait and see what you come up with after investigating fully. A `source` field might be a good fit.

For the tests, you'll have to do whatever makes the most sense to ensure we keep good test coverage. There are two possibilities that come to mind:

* Hard code one of the selection sources, and ensure the code paths for the other selection source are adequately tested elsewhere; or
* Parameterise the test cases over selection source, and run the full suite of tests against both sources.

The latter is likely more effort but also better coverage. It's hard to say which would have a better effort:reward ratio without trying it.

**Comment by vladimir-mencl-eresearch** [comment]
Implemented in #177
